llvm/lib/Transforms/Yk/ControlPoint.cpp

```diff
@@ -158,82 +177,52 @@ void createControlPoint(Module &Mod, Function *F, std::vector<Value *> LiveVars,
-      Mod, Type::getInt8Ty(Context), false, GlobalVariable::InternalLinkage,
-      Int0, "tracing", (GlobalVariable *)nullptr);
-
-  GlobalVariable *GVCompiledTrace = new GlobalVariable(
-      Mod, Type::getInt8PtrTy(Context), false, GlobalVariable::InternalLinkage,
-      PtNull, "compiled_trace", (GlobalVariable *)nullptr);
-
-  GlobalVariable *GVStartLoc = new GlobalVariable(
-      Mod, YkLocTy, false, GlobalVariable::InternalLinkage,
-      Constant::getNullValue(YkLocTy), "start_loc", (GlobalVariable *)nullptr);
-
-  // Create control point entry block. Checks if we are currently tracing.
-  Value *GVTracingVal = Builder.CreateLoad(Type::getInt8Ty(Context), GVTracing);
-  Value *IsTracing =
-      Builder.CreateICmp(CmpInst::Predicate::ICMP_EQ, GVTracingVal, Int0);
-  Builder.CreateCondBr(IsTracing, BBNotTracing, BBTracing);
-
-  // Create block for "not tracing" case. Checks if we already compiled a trace.
-  Builder.SetInsertPoint(BBNotTracing);
-  Value *GVCompiledTraceVal =
-      Builder.CreateLoad(Type::getInt8PtrTy(Context), GVCompiledTrace);
-  Value *HasTrace = Builder.CreateICmp(CmpInst::Predicate::ICMP_EQ,
-                                       GVCompiledTraceVal, PtNull);
-  Builder.CreateCondBr(HasTrace, BBHasNoTrace, BBHasTrace);
-
-  // Create block that starts tracing.
-  Builder.SetInsertPoint(BBHasNoTrace);
+  // Populate the entry block. This calls `__ykrt_transition_location()` to
+  // decide what to do next.
+  IRBuilder<> Builder(CtrlPointEntry);
+  Value *CastLoc =
+      Builder.CreateBitCast(F->getArg(0), Type::getInt8PtrTy(Context));
+  Value *JITAction = Builder.CreateCall(FuncTransLoc->getFunctionType(),
+                                        FuncTransLoc, {CastLoc});
+  SwitchInst *ActionSw = Builder.CreateSwitch(JITAction, BBExecuteTrace, 3);
+  ActionSw->addCase(JActNop, BBReturn);
+  ActionSw->addCase(JActStartTracing, BBStartTracing);
+  ActionSw->addCase(JActStopTracing, BBStopTracing);
+
+  // Populate the block that starts tracing.
+  Builder.SetInsertPoint(BBStartTracing);
   createJITStatePrint(Builder, &Mod, "start-tracing");
   Builder.CreateCall(FuncStartTracing->getFunctionType(), FuncStartTracing,
                      {ConstantInt::get(Context, APInt(64, 1))});
-  Builder.CreateStore(ConstantInt::get(Context, APInt(8, 1)), GVTracing);
-  Builder.CreateStore(F->getArg(0), GVStartLoc);
   Builder.CreateBr(BBReturn);
 
-  // Create block that checks if we've reached the same location again so we
-  // can execute a compiled trace.
-  Builder.SetInsertPoint(BBHasTrace);
-  Value *ValStartLoc = Builder.CreateLoad(YkLocTy, GVStartLoc);
-  Value *ExecTraceCond = Builder.CreateICmp(CmpInst::Predicate::ICMP_EQ,
-                                            ValStartLoc, F->getArg(0));
-  Builder.CreateCondBr(ExecTraceCond, BBExecuteTrace, BBReturn);
-
-  // Create block that executes a compiled trace.
+  // Populate the block that executes a compiled trace. If execution gets into
```

Comment:
> Populate the block that executes a compiled trace.

Does that mean that this block does a call to the generated machine code or ... ?

Reply:
We branch to this block to execute JITted trace code, if `__ykrt_transition_location` says that's what we should do.

I changed "create" to "compiled" since, to be precise, the block was already created, just it's empty.

Does the wording need to be improved?

Reply:
I still don't know what "the block that executes a compiled trace" means! Is that the first block of the trace, or is it a block which then calls the trace?

Reply:
the latter.

Reply:
OK, can we clarify that? I think s/executes/calls/ might be all that's necessary.

Reply:
Fixed in 571fd63e2df7